Theme Showcase: New component SearchThemes

taipeicoder opened the pull request:
#### Proposed Changes

This PR implements a new theme search experience as defined in 0EogtxvlmxMGcnEEaMVECv-fi-1259%3A37689.
![Screen Shot 2022-11-09 at 4 36 17 PM](https://user-images.githubusercontent.com/797888/200780532-80306c9a-4d69-4637-9dac-9be74af13ef7.png)

To accomplish this, we have:
1. Created a new component `<SearchThemes/>` and,
2. Updated the existing component `<KeyedSuggestions/>` to support new UX behaviors.

Note that the `<SearchThemes />` component is based on the existing component `<ThemesMagicSearchCard />` ([src](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/themes/themes-magic-search-card/index.jsx)), and for the most part they should behave similar when it comes to text injection based on suggestion selection. 

The new search experience is as follows:

Whenever the search input field is clicked, the suggestion list will show up and display three main filters: "Feature", "Columns" and "Subject".
![Screen Shot 2022-11-09 at 6 35 33 PM](https://user-images.githubusercontent.com/797888/200808001-b05fcae4-5630-4935-a644-8dbc1cc102a2.png)

Clicking on any of the above will refine the suggestion list to the list of terms from the selected filters. For instance, selecting "Subject" will inject the text `subject:` and update the suggestion list to terms under the filter "Subject", such as "Blog", "Business", etc.
![Screen Shot 2022-11-09 at 6 35 51 PM](https://user-images.githubusercontent.com/797888/200808060-65f3ee3d-0ce6-4903-aa4a-27685fcf1c6c.png)

Clicking on a term will trigger search which updates the theme grid below the search bar. For instance, selecting "Blog" will update the text to `subject:blog` and close the suggestion list. 
![Screen Shot 2022-11-09 at 6 36 19 PM](https://user-images.githubusercontent.com/797888/200808164-d0581a98-f1d6-4410-88e8-ba61b539cae0.png)

Unlike the old search experience, the suggestion list will no longer auto-suggest the first result. This is to allow users to do a plain-text search without being "forced" to select one of the suggestion. To understand more about this, try typing "blog" in the old search experience and do a plain-text search. Note that in the new search experience, keyboard navigation is still available.
![Screen Shot 2022-11-09 at 6 36 44 PM](https://user-images.githubusercontent.com/797888/200808267-bfaf56d2-ac3f-4633-abb6-ef0f2c66775b.png)

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Head to the theme showcase `/themes/${site_slug}`.
  * If using calypso.live, the flag `themes/showcase-i4/search-and-filter` is required.
* Ensure that the search resultset is the same as `wordpress.com/themes/${site_slug}`.
* Ensure that the search and suggestion behavior corresponds to the one described above.
* Ensure that the old search component works as it used to.
  * Disable the new component via flag, i.e.: `/themes/${site_slug}?flags=-themes/showcase-i4/search-and-filter`

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #69658

taipeicoder committed:
New component SearchThemes

arthur791004 commented on client/components/search-themes/index.tsx:
```diff
@@ -0,0 +1,164 @@
+		<div
+			ref={ wrapperRef }
+			className={ classNames( 'search-themes', {
+				'has-keyed-suggestions': isSearchOpen,
```
I guess we don't need this anymore 🤔

taipeicoder replied:
You're right, I've removed it 🙂

arthur791004 commented on client/components/search-themes/style.scss:
```diff
@@ -0,0 +1,170 @@
+	display: flex;
+	align-items: center;
+	background: var(--color-surface);
+	transition: all 0.15s ease-in-out;
```
What's the purpose of this? I cannot see any usage 🤔

taipeicoder replied:
Must have been carried over from the old component 😅 good catch!

arthur791004 commented on client/components/keyed-suggestions/index.jsx:
```diff
@@ -330,28 +373,59 @@ class KeyedSuggestions extends Component {
 	};
 
+	createTopLevelTermsSuggestions = () => {
+		const rendered = [
```
[nit] I think we could return the `JSX.Element` directly and we don't need to follow the `createSuggestions`.

taipeicoder replied:
Good idea, it looks cleaner!

arthur791004 commented on client/components/keyed-suggestions/index.jsx:
```diff
@@ -7,6 +8,34 @@ import { cosineSimilarity } from 'calypso/lib/trigram';
 import './style.scss';
 
+const TAXONOMY_TRANSLATIONS = {
+	feature: i18n.translate( 'Feature', {
```
The translation doesn't work in this way.

![image](https://user-images.githubusercontent.com/13596067/201268919-8261373d-d37a-4edd-8555-a7933efa60b3.png)

We have to define it inside the component or make a function to return the taxonomy as we cannot get locale when we import this file.

taipeicoder replied:
You're right, thanks for pointing this out.

arthur791004 commented on client/components/search-themes/use-outside-click-callback.ts:
```diff
@@ -0,0 +1,44 @@
+ * @param callback   Function to be executed
+ * @param useCapture useCapture flag passed to the add/removeEventListener
+ */
+export default function useOutsideClickCallback(
```
I think we can put it under the `lib/use-outside-click-callback` and share it with `jetpack/profile-dropdown` and her

taipeicoder replied:
I'll create a follow-up issue to have the `jetpack/profile-dropdown` component to import from this file as well (as I'm not sure if it'd trigger code review from other teams, or other deployment processes).

arthur791004 replied:
I found another choice to use the `__experimentalUseFocusOutside` hook from Gutenberg

```js
import { __experimentalUseFocusOutside as useFocusOutside } from '@wordpress/compose';
```

taipeicoder replied:
Changing to `useFocusOutside` seems to work nicely 👍

taipeicoder committed:
Address feedback

taipeicoder committed:
Use __experimentalUseFocusOutside for outside-click detection

arthur791004 reviewed (approved):
Nice job 👍

fushar commented:
Nice! I played around with the feature, and have some feedbacks / clarifications.

1. Looks like there should be a vertical space between the search box and the recommended / trending filter? Currently it looks a bit weird to me as the search box is rounded and the filter below it is not.

<img width="254" alt="image" src="https://user-images.githubusercontent.com/1525580/201631175-f692a830-7a6a-4936-98a3-25ec533da682.png">

fushar commented:
4. Looks like the search box is intended to be sticky. However, it looks like it is composed of two rounded boxes (?), which could make the background to be visible in an awkward way. For example, see the below screenshot: the orange background from the Rainfall theme is "spilling over"

<img width="492" alt="image" src="https://user-images.githubusercontent.com/1525580/201635176-706bc5ea-46bf-4d52-9172-e56a83037ee6.png">

fushar commented:
Other than the UI comments above, the experience feels really good and smooth! I will take a look into the code in more details tomorrow 👍

taipeicoder committed:
Address feedback

taipeicoder commented:
> Looks like there should be a vertical space between the search box and the recommended / trending filter? Currently it looks a bit weird to me as the search box is rounded and the filter below it is not.

I plan to address it in the next PR, which is a revamp of the filter group below 🙂

>When I typed "blog" and hit Enter (to do a plaintext search), the suggestions are not closed and therefore obstruct the (first row of the) search results.

Nice catch! I've fixed it in the latest commit.

>Is the text "all themes" constant? (Havent looked into the code in much details). For example I'm now on Trending themes tab (not all themes) but the placeholder still says "Search all themes", which looks buggy.

You're right, I've changed it to "Search themes..." which is more generic.

>Looks like the search box is intended to be sticky. However, it looks like it is composed of two rounded boxes (?), which could make the background to be visible in an awkward way. For example, see the below screenshot: the orange background from the Rainfall theme is "spilling over"

Fixed, thank you for noticing this 🙏

fushar reviewed (approved):
Looks way better now! 😄

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7625182

Thank you @taipeicoder for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7625182).
